### PR TITLE
Add Tip #141 on Implicit conversions to Bool

### PR DESCRIPTION
**Pull request by manshreck**

### CODE REVIEW

File: _posts/2018-03-15-totw-141.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+This is less clear than the following code:
+
+```c++
+// Better code
+absl::optional<bool> b = MaybeBool();
+if (b.has_value()) { ... }
+```
+
+Note that, in facy, the code snippets above are actually equivalent:
```
Comment (tituswinters) [inline _posts/2018-03-15-totw-141.md]: facy => fact

File: _posts/2018-03-15-totw-141.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+`absl::optional`'s conversion to `bool` only looks at whether the `optional` is
+full, not at its contents. A reader may find it counterintuitive that
```
Comment (tituswinters) [inline _posts/2018-03-15-totw-141.md]: is engaged? has a value?  "is full" seems odd.

File: _posts/2018-03-15-totw-141.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+has a value. Again, it's worth extra caution when the underlying type is
+implicitly convertible to `bool`.
+
+One pattern for optional return-values is to put a variable declaration in the
+conditional of the `if`. This
+[limits the scope](https://google.github.io/styleguide/cppguide.html#Local_Variables)
+of the variable, but involves an implicit conversion to `bool`:
+
+```c++
+if (absl::optional<Foo> foo = MaybeFoo()) {
```
Comment (tituswinters) [inline _posts/2018-03-15-totw-141.md]: Do you want to mention the C++17 variation of this?
https://skebanga.github.io/if-with-initializer/

File: _posts/2018-03-15-totw-141.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+## "Boolean-like" Enums
+
+Let's say you've taken the advice of [TotW #94](/tips/94) and decided to use an
+`enum` in your function signature instead of a `bool` for better readability at
+the call sites. This kind of refactoring might introduce an implicit conversion
+in the function definition:
+
+```c++
+// Bad code
+void InitGoogle(
```
Comment (tituswinters) [inline _posts/2018-03-15-totw-141.md]: The example in #94 externally is ParseCommandLineFlags.

File: _posts/2018-03-15-totw-141.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  }
+}
+```
+
+You can gain additional clarity by replacing the implicit conversion with an
+explicit comparison:
+
+```c++
+// Good code
+void InitGoogle(
```
Comment (tituswinters) [inline _posts/2018-03-15-totw-141.md]: Here as well

File: _posts/2018-03-15-totw-141.md (hunk 1 above)
Comment (manshreck) [reply _posts/2018-03-15-totw-141.md]: Fixed

File: _posts/2018-03-15-totw-141.md (hunk 2 above)
Comment (manshreck) [reply _posts/2018-03-15-totw-141.md]: Yeah, I didn't like "full" either. Changed to " ... whether it contains a value, not at the value itself."

File: _posts/2018-03-15-totw-141.md (hunk 4 above)
Comment (manshreck) [reply _posts/2018-03-15-totw-141.md]: Changed.

File: _posts/2018-03-15-totw-141.md (hunk 5 above)
Comment (manshreck) [reply _posts/2018-03-15-totw-141.md]: Ditto.

File: _posts/2018-03-15-totw-141.md (hunk 3 above)
Comment (manshreck) [reply _posts/2018-03-15-totw-141.md]: Done. Made it inline as an aside so it doesn't detract too much from the main tip.

**Pull request merged**